File: lib/py/uapi/internal/schema/ParseHeadersType.py
from typing import TYPE_CHECKING


if TYPE_CHECKING:
    from uapi.internal.schema.ParseContext import ParseContext
    from uapi.internal.types.UHeaders import UHeaders


def parse_headers_type(path: list[object], headers_definition_as_parsed_json: dict[str, object], schema_key: str, ctx: 'ParseContext') -> 'UHeaders':
    from uapi.UApiSchemaParseError import UApiSchemaParseError
    from uapi.internal.schema.GetTypeUnexpectedParseFailure import get_type_unexpected_parse_failure
    from uapi.internal.types.UFieldDeclaration import UFieldDeclaration
    from uapi.internal.schema.ParseStructFields import parse_struct_fields
    from uapi.internal.schema.SchemaParseFailure import SchemaParseFailure
    from uapi.internal.types.UHeaders import UHeaders

    parse_failures = []
    request_headers = {}
    response_headers = {}

    request_headers_def = headers_definition_as_parsed_json.get(schema_key)

    this_path = path + [schema_key]

    if not isinstance(request_headers_def, dict):
        branch_parse_failures = get_type_unexpected_parse_failure(
            ctx.document_name,
            this_path,
            request_headers_def,
            'Object',
        )
        parse_failures.extend(branch_parse_failures)
    else:
        try:
            request_fields = parse_struct_fields(this_path, request_headers_def, ctx)

            # All headers are optional
            for field in request_fields:
                request_fields[field].optional = True

            request_headers.update(request_fields)
        except UApiSchemaParseError as e:
            parse_failures.extend(e.schema_parse_failures)
        except Exception as e:
            raise e

    response_key = '->'
    response_path = path + [response_key]

    if response_key not in headers_definition_as_parsed_json:
        parse_failures.append(
            SchemaParseFailure(ctx.document_name, response_path, 'RequiredObjectKeyMissing', {
                'key': response_key,
            })
        )

    response_headers_def = headers_definition_as_parsed_json.get(response_key)

    if not isinstance(response_headers_def, dict):
        branch_parse_failures = get_type_unexpected_parse_failure(
            ctx.document_name,
            this_path,
            response_headers_def,
            'Object',
        )
        parse_failures.extend(branch_parse_failures)
    else:
        try:
            response_fields = parse_struct_fields(response_path, response_headers_def, ctx)

            # All headers are optional
            for field in response_fields:
                response_fields[field].optional = True

            response_headers.update(response_fields)
        except UApiSchemaParseError as e:
            parse_failures.extend(e.schema_parse_failures)
        except Exception as e:
            raise e

    if parse_failures:
        raise UApiSchemaParseError(parse_failures, ctx.uapi_schema_document_names_to_json)

    return UHeaders(schema_key, request_headers, response_headers)

File: lib/java/src/main/java/uapi/internal/schema/ParseHeadersType.java
package uapi.internal.schema;

import uapi.internal.types.UFieldDeclaration;
import uapi.internal.types.UHeaders;
import uapi.UApiSchemaParseError;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static uapi.internal.schema.ParseStructFields.parseStructFields;

public class ParseHeadersType {

    public static UHeaders parseHeadersType(
            List<Object> path,
            Map<String, Object> headersDefinitionAsParsedJson,
            String schemaKey,
            ParseContext ctx) throws UApiSchemaParseError {
        List<SchemaParseFailure> parseFailures = new ArrayList<>();
        Map<String, UFieldDeclaration> requestHeaders = new HashMap<>();
        Map<String, UFieldDeclaration> responseHeaders = new HashMap<>();

        Object requestHeadersDef = headersDefinitionAsParsedJson.get(schemaKey);

        List<Object> thisPath = new ArrayList<>(path);
        thisPath.add(schemaKey);

        if (!(requestHeadersDef instanceof Map)) {
            List<SchemaParseFailure> branchParseFailures = GetTypeUnexpectedParseFailure.getTypeUnexpectedParseFailure(
                    ctx.documentName,
                    thisPath,
                    requestHeadersDef,
                    "Object");
            parseFailures.addAll(branchParseFailures);
        } else {
            try {
                Map<String, UFieldDeclaration> requestFields = parseStructFields(thisPath,
                        (Map<String, Object>) requestHeadersDef, ctx);

                // All headers are optional
                final var finalRequestFields = requestFields.entrySet().stream()
                        .collect(Collectors.toMap(e -> e.getKey(), e -> new UFieldDeclaration(e.getValue().fieldName,
                                e.getValue().typeDeclaration, true)));

                requestHeaders.putAll(finalRequestFields);
            } catch (UApiSchemaParseError e) {
                parseFailures.addAll(e.schemaParseFailures);
            }
        }

        String responseKey = "->";
        List<Object> responsePath = new ArrayList<>(path);
        responsePath.add(responseKey);

        if (!headersDefinitionAsParsedJson.containsKey(responseKey)) {
            parseFailures.add(
                    new SchemaParseFailure(ctx.documentName, responsePath, "RequiredObjectKeyMissing",
                            Map.of("key", responseKey)));
        }

        Object responseHeadersDef = headersDefinitionAsParsedJson.get(responseKey);

        if (!(responseHeadersDef instanceof Map)) {
            List<SchemaParseFailure> branchParseFailures = GetTypeUnexpectedParseFailure.getTypeUnexpectedParseFailure(
                    ctx.documentName,
                    thisPath,
                    responseHeadersDef,
                    "Object");
            parseFailures.addAll(branchParseFailures);
        } else {
            try {
                Map<String, UFieldDeclaration> responseFields = ParseStructFields.parseStructFields(responsePath,
                        (Map<String, Object>) responseHeadersDef, ctx);

                // All headers are optional
                final var finalResponseFields = responseFields.entrySet().stream()
                        .collect(Collectors.toMap(e -> e.getKey(), e -> new UFieldDeclaration(e.getValue().fieldName,
                                e.getValue().typeDeclaration, true)));

                responseHeaders.putAll(finalResponseFields);
            } catch (UApiSchemaParseError e) {
                parseFailures.addAll(e.schemaParseFailures);
            }
        }

        if (!parseFailures.isEmpty()) {
            throw new UApiSchemaParseError(parseFailures, ctx.uApiSchemaDocumentNamesToJson);
        }

        return new UHeaders(schemaKey, requestHeaders, responseHeaders);
    }
}
